hibernate-core/src/main/java/org/hibernate/dialect/function/CommonFunctionFactory.java

```diff
@@ -1476,7 +1490,83 @@ public static void aggregates(QueryEngine queryEngine) {
+		final BasicType<Long> longType = typeConfiguration.getBasicTypeForJavaType( Long.class );
+		final BasicType<Double> doubleType = typeConfiguration.getBasicTypeForJavaType( Double.class );
+		final BasicType<BigInteger> bigIntegerType = typeConfiguration.getBasicTypeForJavaType( BigInteger.class );
+		final BasicType<BigDecimal> bigDecimalType = typeConfiguration.getBasicTypeForJavaType( BigDecimal.class );
 		queryEngine.getSqmFunctionRegistry().namedDescriptorBuilder("sum")
+				.setReturnTypeResolver( new FunctionReturnTypeResolver() {
+					@Override
+					public AllowableFunctionReturnType<?> resolveFunctionReturnType(AllowableFunctionReturnType<?> impliedType, List<SqmTypedNode<?>> arguments, TypeConfiguration typeConfiguration) {
+						final AllowableFunctionReturnType<?> argType = StandardFunctionReturnTypeResolvers.extractArgumentType(
+								arguments,
+								1
+						);
+						final BasicType<?> basicType;
+						if ( argType instanceof BasicType<?> ) {
+							basicType = (BasicType<?>) argType;
+						}
+						else {
+							basicType = typeConfiguration.getBasicTypeForJavaType( argType.getJavaType() );
+							if ( basicType == null ) {
+								return impliedType;
+							}
+						}
+						// Resolve according to JPA spec 4.8.5
+						switch ( basicType.getSqlTypeDescriptor().getJdbcTypeCode() ) {
+							case Types.SMALLINT:
+							case Types.TINYINT:
+							case Types.INTEGER:
+							case Types.BIGINT:
+								return longType;
+							case Types.FLOAT:
+							case Types.REAL:
+							case Types.DOUBLE:
+								return doubleType;
+							case Types.DECIMAL:
+							case Types.NUMERIC:
+								if ( BigInteger.class.isAssignableFrom( basicType.getJavaType() ) ) {
+									return bigIntegerType;
+								}
+								else {
+									return bigDecimalType;
+								}
+						}
+						// Better use the implied type than throwing an exception
+						return impliedType;
+					}
+
+					@Override
+					public BasicValuedMapping resolveFunctionReturnType(Supplier<BasicValuedMapping> impliedTypeAccess, List<? extends SqlAstNode> arguments) {
+						// Resolve according to JPA spec 4.8.5
+						final BasicValuedMapping specifiedArgType = StandardFunctionReturnTypeResolvers.extractArgumentValuedMapping(
+								arguments,
+								1
+						);
+						switch ( specifiedArgType.getJdbcMapping().getSqlTypeDescriptor().getJdbcTypeCode() ) {
+							case Types.SMALLINT:
+							case Types.TINYINT:
+							case Types.INTEGER:
+							case Types.BIGINT:
+								return longType;
+							case Types.FLOAT:
+							case Types.REAL:
+							case Types.DOUBLE:
+								return doubleType;
+							case Types.DECIMAL:
+							case Types.NUMERIC:
+								if ( BigInteger.class.isAssignableFrom( specifiedArgType.getJdbcMapping().getJavaTypeDescriptor().getJavaTypeClass() ) ) {
+									return bigIntegerType;
+								}
+								else {
+									return bigDecimalType;
+								}
+						}
+						return impliedTypeAccess.get();
+					}
+
+				} )
```

Comment:
Good catch! 

I did not notice that the JPA spec (Section 4.8.5) is not silent on this issue:

> SUM returns Long when applied to state fields of integral types (other than BigInteger); Double when applied to state fields of floating point types; BigInteger when applied to state fields of type BigInteger; and BigDecimal when applied to state fields of type BigDecimal.

Could you do me a favor and copy that spec language into a comment, so we know why this code is necessary.